Type description

oliverguenther opened the pull request.

oliverguenther committed:
Add description to type defaults

ulferts commented on lib/api/v3/work_packages/work_package_representer.rb:
```diff
@@ -324,7 +324,14 @@ def initialize(model, current_user:, embed_links: false)
+                     ::API::Decorators::Formattable.new(type.description, object: represented)
+                   else
+                     ::API::Decorators::Formattable.new(description, object: represented)
+                   end
```
If I am not mistaken, this will lead to all work packages not having a description receiving the `type.description` (if that is defined) even if they are already persisted which is not what is intended, I believe.

Additionally, I think that this kind of business logic does not belong into the representer, but rather into the service layer.

ulferts replied:
https://github.com/opf/openproject/blob/dev/app/services/work_packages/set_attributes_service.rb#L85 is a good place

oliverguenther committed:
Use SetAttributeService

HDinger committed:
Changing the type afterwards leads to a change in the description if the description is empty or still the default template. Changes made by the user shall not be overridden.

HDinger reviewed (commented):
There was an issue when changing the type later in the creation process (Click on "Create --> Bug", change the type then). However when I did not touch the description I expect it to change accordingly. 

We added a check in the backend that should set the default_description when the field is either empty or matches one of the templates. I suggest to shortly discuss this, when your back @oliverguenther

oliverguenther commented:
This is due to the `set_default_attributes` method is the wrong place, because type is only set afterwards in `update_project_dependent_attributes`. There is another problem when uploading attachments because the form then returns an array of the attachment IDs and this will break uploading further attachments when the form payload gets `$initialize`d back into the work package. I will need to think about how this could work without modifying the work package.

oliverguenther committed:
Correctly reset frontend when description changed

This requires the frontend to know that description is a field we take
over from the form without changing the underlying work package.

Since we cannot blindly take over all values from the from due to
inconsistencies in attachment links, the frontend needs to know.
